crates/ruff_python_formatter/src/string/normalize.rs

```diff
@@ -73,18 +74,23 @@ impl StringNormalizer {
     }
 
     /// Computes the strings preferred quotes.
-    pub(crate) fn choose_quotes(&self, string: &StringPart, locator: &Locator) -> QuoteSelection {
+    pub(crate) fn choose_quotes(
+        &self,
+        string: &impl StringPart,
```

Comment:
Using `impl` here means that all these methods get monomorphized for each `StringPart` type. It's not a ton of code but still considerable. 


The other downside is that `StringPart` is a type from the `ruff_python_parser` crate. Ideally, the formatter would only depend on the AST, but not the parser and would instead be given a parsed AST already (the formatter doesn't really care about parsing). 



That's why I think I would prefer keeping the `StringPart` enum (or you can have another struct that holds the relevant field and implements `From` methods for all string types). E.g. all we need is the `kind`, and `possibly` the range. So you could define a `StringPart` struct that stores `kind` and `range` and implements `From` for all possible string part values. This way you avoid monomorphization and depending on a rust parser crate type.